File: etapa3-construcao(geracao-de-cod)/stack2/src/domain/entities/Quarto.ts
import { TipoQuarto, Disponibilidade, ValidationResult, QuartoData } from '../../types';
import { Cama } from './Cama';
import { v4 as uuidv4 } from 'uuid';

export class Quarto {
  readonly id: string;
  private _numero: number;
  private _capacidade: number;
  private _tipo: TipoQuarto;
  private _precoPorDiaria: number;
  private _temFrigobar: boolean;
  private _temCafeDaManha: boolean;
  private _temArCondicionado: boolean;
  private _temTV: boolean;
  private _disponibilidade: Disponibilidade;
  private _camas: Cama[];
  readonly criadoEm: Date;
  private _atualizadoEm: Date;

  constructor(params: {
    numero: number;
    capacidade: number;
    tipo: TipoQuarto;
    precoPorDiaria: number;
    temFrigobar: boolean;
    temCafeDaManha: boolean;
    temArCondicionado: boolean;
    temTV: boolean;
    disponibilidade?: Disponibilidade;
    camas?: Cama[];
    id?: string;
    criadoEm?: Date;
    atualizadoEm?: Date;
  }) {
    this.id = params.id || uuidv4();
    this._numero = params.numero;
    this._capacidade = params.capacidade;
    this._tipo = params.tipo;
    this._precoPorDiaria = params.precoPorDiaria;
    this._temFrigobar = params.temFrigobar;
    this._temCafeDaManha = params.temCafeDaManha;
    this._temArCondicionado = params.temArCondicionado;
    this._temTV = params.temTV;
    this._disponibilidade = params.disponibilidade || Disponibilidade.LIVRE;
    this._camas = params.camas || [];
    this.criadoEm = params.criadoEm || new Date();
    this._atualizadoEm = params.atualizadoEm || new Date();
  }

  get numero(): number {
    return this._numero;
  }

  get capacidade(): number {
    return this._capacidade;
  }

  get tipo(): TipoQuarto {
    return this._tipo;
  }

  get precoPorDiaria(): number {
    return this._precoPorDiaria;
  }

  get temFrigobar(): boolean {
    return this._temFrigobar;
  }

  get temCafeDaManha(): boolean {
    return this._temCafeDaManha;
  }

  get temArCondicionado(): boolean {
    return this._temArCondicionado;
  }

  get temTV(): boolean {
    return this._temTV;
  }

  get disponibilidade(): Disponibilidade {
    return this._disponibilidade;
  }

  get camas(): Cama[] {
    return [...this._camas];
  }

  get atualizadoEm(): Date {
    return this._atualizadoEm;
  }

  alterarDisponibilidade(status: Disponibilidade): void {
    this._disponibilidade = status;
    this._atualizadoEm = new Date();
  }

  podeSerReservado(): boolean {
    return this._disponibilidade === Disponibilidade.LIVRE;
  }

  adicionarCama(cama: Cama): void {
    this._camas.push(cama);
    this._atualizadoEm = new Date();
  }

  removerCama(camaId: string): void {
    this._camas = this._camas.filter(c => c.id !== camaId);
    this._atualizadoEm = new Date();
  }

  atualizarDados(params: {
    numero?: number;
    capacidade?: number;
    tipo?: TipoQuarto;
    precoPorDiaria?: number;
    temFrigobar?: boolean;
    temCafeDaManha?: boolean;
    temArCondicionado?: boolean;
    temTV?: boolean;
    camas?: Cama[];
  }): void {
    if (params.numero !== undefined) this._numero = params.numero;
    if (params.capacidade !== undefined) this._capacidade = params.capacidade;
    if (params.tipo !== undefined) this._tipo = params.tipo;
    if (params.precoPorDiaria !== undefined) this._precoPorDiaria = params.precoPorDiaria;
    if (params.temFrigobar !== undefined) this._temFrigobar = params.temFrigobar;
    if (params.temCafeDaManha !== undefined) this._temCafeDaManha = params.temCafeDaManha;
    if (params.temArCondicionado !== undefined) this._temArCondicionado = params.temArCondicionado;
    if (params.temTV !== undefined) this._temTV = params.temTV;
    if (params.camas !== undefined) this._camas = params.camas;
    
    this._atualizadoEm = new Date();
  }

  calcularValorTotal(numeroDiarias: number): number {
    return this._precoPorDiaria * numeroDiarias;
  }

  validar(): ValidationResult {
    const errors: string[] = [];

    if (this._numero <= 0) {
      errors.push('Número do quarto deve ser maior que zero');
    }

    if (this._capacidade <= 0) {
      errors.push('Capacidade deve ser maior que zero');
    }

    if (this._precoPorDiaria <= 0) {
      errors.push('Preço por diária deve ser maior que zero');
    }

    if (this._camas.length === 0) {
      errors.push('Quarto deve ter pelo menos uma cama');
    }

    return {
      isValid: errors.length === 0,
      errors
    };
  }

  toData(): QuartoData {
    return {
      id: this.id,
      numero: this._numero,
      capacidade: this._capacidade,
      tipo: this._tipo,
      precoPorDiaria: this._precoPorDiaria,
      temFrigobar: this._temFrigobar,
      temCafeDaManha: this._temCafeDaManha,
      temArCondicionado: this._temArCondicionado,
      temTV: this._temTV,
      disponibilidade: this._disponibilidade,
      camas: this._camas.map(c => c.toData()),
      criadoEm: this.criadoEm,
      atualizadoEm: this._atualizadoEm
    };
  }

  static fromData(data: QuartoData): Quarto {
    return new Quarto({
      numero: data.numero,
      capacidade: data.capacidade,
      tipo: data.tipo,
      precoPorDiaria: data.precoPorDiaria,
      temFrigobar: data.temFrigobar,
      temCafeDaManha: data.temCafeDaManha,
      temArCondicionado: data.temArCondicionado,
      temTV: data.temTV,
      disponibilidade: data.disponibilidade,
      camas: data.camas.map(c => Cama.fromData(c)),
      id: data.id,
      criadoEm: data.criadoEm,
      atualizadoEm: data.atualizadoEm
    });
  }
}

File: etapa3-construcao(geracao-de-cod)/stack2/src/backend/src/main/java/com/hotel/reserva/domain/entities/Quarto.java
package com.hotel.reserva.domain.entities;

import com.hotel.reserva.domain.enums.Disponibilidade;
import com.hotel.reserva.domain.enums.TipoCama;
import com.hotel.reserva.domain.enums.TipoQuarto;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "quartos")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@EqualsAndHashCode(of = "id")
public class Quarto {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false, unique = true)
    private Integer numero;

    @Column(nullable = false)
    private Integer capacidade;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TipoQuarto tipo;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal precoPorDiaria;

    @Column(nullable = false)
    private Boolean temFrigobar;

    @Column(nullable = false)
    private Boolean temCafeDaManha;

    @Column(nullable = false)
    private Boolean temArCondicionado;

    @Column(nullable = false)
    private Boolean temTV;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Disponibilidade disponibilidade;

    @OneToMany(mappedBy = "quarto", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Cama> camas = new ArrayList<>();

    @Column(nullable = false, updatable = false)
    private LocalDateTime criadoEm;

    @Column(nullable = false)
    private LocalDateTime atualizadoEm;

    @PrePersist
    protected void onCreate() {
        criadoEm = LocalDateTime.now();
        atualizadoEm = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        atualizadoEm = LocalDateTime.now();
    }

    @Builder
    public Quarto(Integer numero, Integer capacidade, TipoQuarto tipo,
                  BigDecimal precoPorDiaria, Boolean temFrigobar, Boolean temCafeDaManha,
                  Boolean temArCondicionado, Boolean temTV, List<TipoCama> tiposCamas) {
        this.numero = numero;
        this.capacidade = capacidade;
        this.tipo = tipo;
        this.precoPorDiaria = precoPorDiaria;
        this.temFrigobar = temFrigobar;
        this.temCafeDaManha = temCafeDaManha;
        this.temArCondicionado = temArCondicionado;
        this.temTV = temTV;
        this.disponibilidade = Disponibilidade.LIVRE;
        
        if (tiposCamas != null) {
            tiposCamas.forEach(this::adicionarCama);
        }
    }

    public void alterarDisponibilidade(Disponibilidade novaDisponibilidade) {
        this.disponibilidade = novaDisponibilidade;
        this.atualizadoEm = LocalDateTime.now();
    }

    public boolean podeSerReservado() {
        return this.disponibilidade == Disponibilidade.LIVRE;
    }

    public void adicionarCama(TipoCama tipo) {
        Cama cama = new Cama(tipo);
        cama.setQuarto(this);
        this.camas.add(cama);
    }

    public void removerCama(UUID camaId) {
        this.camas.removeIf(cama -> cama.getId().equals(camaId));
    }

    public void atualizarDados(Integer numero, Integer capacidade, TipoQuarto tipo,
                              BigDecimal precoPorDiaria, Boolean temFrigobar, Boolean temCafeDaManha,
                              Boolean temArCondicionado, Boolean temTV, List<TipoCama> tiposCamas) {
        if (numero != null) this.numero = numero;
        if (capacidade != null) this.capacidade = capacidade;
        if (tipo != null) this.tipo = tipo;
        if (precoPorDiaria != null) this.precoPorDiaria = precoPorDiaria;
        if (temFrigobar != null) this.temFrigobar = temFrigobar;
        if (temCafeDaManha != null) this.temCafeDaManha = temCafeDaManha;
        if (temArCondicionado != null) this.temArCondicionado = temArCondicionado;
        if (temTV != null) this.temTV = temTV;
        
        if (tiposCamas != null) {
            this.camas.clear();
            tiposCamas.forEach(this::adicionarCama);
        }
        
        this.atualizadoEm = LocalDateTime.now();
    }

    public BigDecimal calcularValorTotal(int numeroDiarias) {
        return precoPorDiaria.multiply(BigDecimal.valueOf(numeroDiarias));
    }

    public List<Cama> getCamas() {
        return Collections.unmodifiableList(camas);
    }

    public void validar() {
        if (numero == null || numero <= 0) {
            throw new IllegalArgumentException("Número do quarto deve ser maior que zero");
        }
        
        if (capacidade == null || capacidade <= 0) {
            throw new IllegalArgumentException("Capacidade deve ser maior que zero");
        }
        
        if (precoPorDiaria == null || precoPorDiaria.compareTo(BigDecimal.ZERO) <= 0) {
            throw new IllegalArgumentException("Preço por diária deve ser maior que zero");
        }
        
        if (camas.isEmpty()) {
            throw new IllegalArgumentException("Quarto deve ter pelo menos uma cama");
        }
    }
}
